tests/utilities/test_auto_restart.py

```diff
@@ -893,3 +903,59 @@ def create_dataloader():
 
     state_dict = dataloader_to_state_dict(dataloader, iter_dataloader)
     assert state_dict == {'num_workers': 0, 'previous_worker': None, 0: {'current_iteration': 24}}
+
+
+@pytest.mark.parametrize("use_fault_tolerant", ['0', '1'])
+@RunIf(min_torch="1.7.0")
+def test_data_loading_wraps_dataset_and_samplers(use_fault_tolerant, tmpdir):
+    """
+    this test ensures the dataset and sampler are properly wrapped when fault tolerant is enabled.
+    """
+
+    class CustomBatchSampler(BatchSampler):
+        pass
+
+    dataset = range(50)
+
+    class TestModel(BoringModel):
+
+        def train_dataloader(self):
+            return {
+                "a": [
+                    DataLoader(create_iterable_dataset(3, 1, wrap=False), num_workers=0, batch_size=3),
+                    DataLoader(dataset, batch_size=8),
+                    DataLoader(
+                        dataset,
+                        batch_sampler=CustomBatchSampler(SequentialSampler(dataset), batch_size=8, drop_last=False)
+                    ),
+                ],
+                "b": DataLoader(
+                    create_iterable_dataset(2, num_workers=1, attr_name="custom_sampler", wrap=False),
+                    num_workers=0,
+                    batch_size=2
+                )
+            }
+
+        def training_step(self, batch, batch_idx):
+            pass
+
+    class Check(Callback):
+
+        def on_train_batch_start(self, trainer, *_) -> None:
+            loaders = trainer.train_dataloader.loaders
+            if use_fault_tolerant == '1':
+                assert isinstance(loaders["a"][0].loader.dataset, CaptureIterableDataset)
+                assert isinstance(loaders["a"][1].loader.sampler, FastForwardSampler)
+                assert isinstance(loaders["a"][2].loader.batch_sampler, FastForwardSampler)
+                assert isinstance(loaders["b"].loader.dataset, CaptureIterableDataset)
+            else:
+                assert isinstance(loaders["a"][0].loader.dataset, RangeIterableDataset)
+                assert isinstance(loaders["a"][1].loader.sampler, SequentialSampler)
+                assert isinstance(loaders["a"][2].loader.batch_sampler, CustomBatchSampler)
+                assert isinstance(loaders["b"].loader.dataset, RangeIterableDataset)
```

Comment:
Not sure if these are worth adding considering they will be removed soon

Reply:
Why would they be removed ?

Reply:
I mean when the fault tolerance experimental flag is removed :)